### PR TITLE
fix: bitswap query timeout

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

Changes introduced in this pull request:
- Call `cancel` if `bitswap` requests are not responded to after 5s
- Update changelog, adding `fixed` section

Feature is not user-facing

[CI log](https://github.com/ChainSafe/forest/actions/runs/3882657409/jobs/6623063511#step:13:1754) looks correct:
```console
libp2p_messsage_total{libp2p_message_kind="bitswap_block_request_out"} 21
libp2p_messsage_total{libp2p_message_kind="bitswap_block_response_in"} 19
...
network_container_capacities{kind="bitswap_outgoing_query_ids"} 14
network_container_capacities{kind="cx_request_table"} 7
network_container_capacities{kind="hello_request_table"} 14
```

## Reference issue to close (if applicable)

Closes https://github.com/ChainSafe/forest/issues/2398

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG](https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md) is up-to-date. All user-facing changes should be reflected in this document.

### CODE REVIEW

**Comment by elmattic** [comment]
Not sure if already ready but looks like I have some stall when syncing (ctrl-c forest after 5 minutes):

```
./target/release/forest --encrypt-keystore false --target-peer-count 50
2023-01-09T19:05:57.323019Z  INFO forest: Using default mainnet config    
2023-01-09T19:05:57.323806Z  INFO forest::daemon: Starting Forest daemon, version 0.4.1+git.0b76e4c5    
2023-01-09T19:05:57.324126Z  INFO forest_libp2p::service: Recovered libp2p keypair from "/Users/guillaume/Library/Application Support/com.ChainSafe.Forest/libp2p/keypair"    
2023-01-09T19:05:57.324176Z  INFO forest::daemon: PeerId: 12D3KooWQGY99GfNXhHRCoeUnU4cRK9T9TsZA9vpxbqh6gGXmjSp    
2023-01-09T19:05:57.324182Z  WARN forest::daemon: Warning: Keystore encryption disabled!    
2023-01-09T19:05:57.324612Z  INFO forest::daemon: Admin token: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXSwiZXhwIjoxNjc4NDc1MTU3fQ.OdpOVsRgr9cfnc0PgMnm56PXNchzNYQAfaw1nbBJjp0    
2023-01-09T19:05:57.363770Z  INFO forest::daemon: Prometheus server started at 0.0.0.0:6116    
2023-01-09T19:05:57.370888Z  INFO forest_genesis: Initialized genesis: BlockHeader: Cid(bafy2bzacecnamqgqmifpluoeldx7zzglxcljo6oja4vrmtj7432rphldpdmm2)    
2023-01-09T19:05:57.371036Z  INFO forest::daemon: Using network :: mainnet    
2023-01-09T19:05:57.373733Z  INFO forest_chain_sync::chain_muxer: Evaluating network head...    
2023-01-09T19:05:57.373837Z  INFO forest::daemon: JSON-RPC endpoint started at 127.0.0.1:1234    
2023-01-09T19:05:57.374028Z  INFO forest_rpc: Ready for RPC connections    
2023-01-09T19:05:57.710737Z  INFO forest_libp2p::service: Running libp2p service    
2023-01-09T19:05:58.828298Z  INFO forest_chain_sync::chain_muxer: Local node is behind the network, starting BOOTSTRAP from LOCAL_HEAD = 2499415 -> NETWORK_HEAD = 2499491    
Downloading headers 76 / 76 [====================================================================================================================================] 100.00 % 1030.58/s 2023-01-09T19:05:59.005508Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2499415, N blocks = 2    
2023-01-09T19:06:08.933515Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2499416, N blocks = 4    
2023-01-09T19:06:12.091095Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2499417, N blocks = 5    
2023-01-09T19:06:18.812607Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2499418, N blocks = 5    
2023-01-09T19:06:23.405985Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2499419, N blocks = 6    
2023-01-09T19:06:25.391653Z  WARN forest_libp2p::service: RPCResponse receive failed    
2023-01-09T19:06:37.040348Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2499420, N blocks = 11    
2023-01-09T19:06:42.816320Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2499421, N blocks = 4    
2023-01-09T19:06:47.358377Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2499422, N blocks = 3    
2023-01-09T19:06:50.972237Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2499423, N blocks = 2    
2023-01-09T19:06:54.735976Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2499424, N blocks = 5    
2023-01-09T19:06:58.808168Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2499425, N blocks = 5    
2023-01-09T19:07:04.600038Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2499426, N blocks = 1    
2023-01-09T19:07:08.119494Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2499427, N blocks = 3    
2023-01-09T19:07:11.648083Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2499428, N blocks = 6    
2023-01-09T19:07:15.443440Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2499430, N blocks = 2    
2023-01-09T19:07:22.537771Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2499431, N blocks = 5    
2023-01-09T19:07:52.387728Z ERROR forest_chain_sync::tipset_syncer: Sync messages check state failed for tipset range    
2023-01-09T19:07:52.387822Z ERROR forest_chain_sync::chain_muxer: Bootstrapping failed, re-evaluating the network head to retry the bootstrap. Error = TipsetRangeSyncer(NetworkMessageQueryFailed("ChainExchange request failed for all top peers. 26 network failures, 0 lookup failures, request:\nChainExchangeRequest { start: [Cid(bafy2bzacecwbb5qfvfeczfgehoaemnf3zgvbfa4veiljxxj3ug25dhblyhqyq), Cid(bafy2bzacedtmxupghl6m7prauc7m4gcc6a5vkntg6nasoomcpzvpdsuvsp5ps), Cid(bafy2bzaceanspck2ialpeqgud4ffqh7pthtyiyysxw4mflpfdx6i6ktffveiq), Cid(bafy2bzaceda5rpns545fq2buycixrsgw5dsdtko6nkxo4wfalsettti25y5he), Cid(bafy2bzaceb3us2frwohd6rc6ljhfs6yqj7uc6vzxsiaxxfqyvlcecyjaohvp4)], request_len: 4, options: 2 }"))    
2023-01-09T19:07:52.387913Z  INFO forest_chain_sync::chain_muxer: Evaluating network head...    

^C2023-01-09T19:12:12.788345Z  WARN forest::cli: Got interrupt, shutting down...    
2023-01-09T19:12:12.788941Z  WARN forest_libp2p::service: bitswap query cancellation err: sending on a closed channel    
2023-01-09T19:12:12.788960Z  WARN forest_libp2p::service: bitswap query cancellation err: sending on a closed channel    
2023-01-09T19:12:12.788969Z  WARN forest_libp2p::service: bitswap query cancellation err: sending on a closed channe
```l

**Comment by hanabi1224** [comment]
@elmattic Thanks for testing! It's caused by a deadlock, trying to read the lock while holding the write lock, now fixed
```rust
if let Some((cid, _)) = outgoing_bitswap_query_ids.write().await.remove(&query_id) {
                    metrics::NETWORK_CONTAINER_CAPACITIES
                        .with_label_values(&[metrics::values::BITSWAP_OUTGOING_QUERY_IDS])
                        .set(outgoing_bitswap_query_ids.read().await.capacity() as u64);
...
}
```

File: node/forest_libp2p/src/service.rs [hunk 1]
```diff
@@ -342,13 +361,48 @@ where
...
+                    locked.remove(id);
+                }
+            }
+            for id in timeout_queries {
+                if let Err(e) = outgoing_bitswap_query_cancellation_tx.send_async(id).await {
+                    warn!("bitswap query cancellation err: {e}");
+                }
+            }
+        }
+        metrics::NETWORK_CONTAINER_CAPACITIES
```
Comment (hanabi1224) [inline node/forest_libp2p/src/service.rs:391]: Moved capacity tracking logic here to simplify code and avoid careless deadlock

**Comment by elmattic** [comment]
I would demote those info logs, else it will a bit noisy:

```
2023-01-10T12:09:39.897969Z  WARN forest_libp2p::service: RPCResponse receive failed    
2023-01-10T12:09:40.719572Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2501533, N blocks = 7    
2023-01-10T12:09:41.547254Z  WARN forest_libp2p::service: RPCResponse receive failed    
2023-01-10T12:09:44.273600Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2501534, N blocks = 6    
2023-01-10T12:09:48.305293Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2501535, N blocks = 4    
2023-01-10T12:09:52.011478Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2501536, N blocks = 4    
2023-01-10T12:10:02.388090Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2501537, N blocks = 5    
2023-01-10T12:10:07.261786Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 2501538, N blocks = 4    
2023-01-10T12:10:10.836646Z  INFO forest_chain::store::chain_store: New heaviest tipset: TipsetKeys { cids: [Cid(bafy2bzacec7nhxp4bhvcur62tgflqfzyruixasspbtvqf5z5k4rvoxfdlbqwa), Cid(bafy2bzacecm3kv4sqrkmyd5nb56jcxzvcpf4pbjt76bfqbyuhja52xworykxa), Cid(bafy2bzaceanvxvll5uik7jajefsidlsipxh5k5xl3w6tnkfbkoc3hdadiv472), Cid(bafy2bzacebu5ppmacnkqpvi3ll6ezvsas5mtm2qzmrjgq4kh4dy5qrmbkum64)] }    
2023-01-10T12:10:10.836678Z  INFO forest_chain_sync::chain_muxer: Bootstrap successfully completed, now evaluating the network head to ensure the node is in sync    
2023-01-10T12:10:10.836680Z  INFO forest_chain_sync::chain_muxer: Evaluating network head...    
2023-01-10T12:10:15.711986Z  INFO forest_libp2p::service: Cancelling bitswap query 1134    
2023-01-10T12:10:15.712082Z  INFO forest_libp2p::service: Cancelling bitswap query 19708    
2023-01-10T12:10:15.712113Z  INFO forest_libp2p::service: Cancelling bitswap query 21457    
2023-01-10T12:10:15.712138Z  INFO forest_libp2p::service: Cancelling bitswap query 8046    
2023-01-10T12:10:15.712163Z  INFO forest_libp2p::service: Cancelling bitswap query 27340    
2023-01-10T12:10:15.712181Z  INFO forest_libp2p::service: Cancelling bitswap query 17323    
2023-01-10T12:10:15.712208Z  INFO forest_libp2p::service: Cancelling bitswap query 27658    
2023-01-10T12:10:15.712229Z  INFO forest_libp2p::service: Cancelling bitswap query 28506    
2023-01-10T12:10:15.712247Z  INFO forest_libp2p::service: Cancelling bitswap query 10910    
2023-01-10T12:10:15.712265Z  INFO forest_libp2p::service: Cancelling bitswap query 28612    
2023-01-10T12:10:15.712281Z  INFO forest_libp2p::service: Cancelling bitswap query 29301    
2023-01-10T12:10:15.712304Z  INFO forest_libp2p::service: Cancelling bitswap query 29937    
2023-01-10T12:10:15.712331Z  INFO forest_libp2p::service: Cancelling bitswap query 2052    
2023-01-10T12:10:15.712352Z  INFO forest_libp2p::service: Cancelling bitswap query 22358    
2023-01-10T12:10:15.712370Z  INFO forest_libp2p::service: Cancelling bitswap query 22411 
```

**Comment by hanabi1224** [comment]
> I would demote those info logs, else it will a bit noisy:

@elmattic fixed

**Review by lemmih — APPROVED** [review]

**Review by tyshko5 — APPROVED** [review]

**Comment by hanabi1224** [comment]
> @hanabi1224 What about internal containers size in `bitswap` crate?

By reading the code I think it should be fixed by this PR, maybe I can add some metrics to my fork if that makes ur testing easier

**Comment by elmattic** [comment]
> > @hanabi1224 What about internal containers size in `bitswap` crate?
> 
> By reading the code I think it should be fixed by this PR, maybe I can add some metrics to my fork if that makes ur testing easier

That's ok, I have some logs in my local fork as well, so I can check.

**Comment by elmattic** [comment]
> > @hanabi1224 What about internal containers size in `bitswap` crate?
> 
> By reading the code I think it should be fixed by this PR, maybe I can add some metrics to my fork if that makes ur testing easier

Yes it would help to add those metrics, just to be sure.
I did some testing and I can see capacity of `queries: FnvHashMap<QueryId, Query>` increasing like this over a period of 90 minutes:
- 457,443
- 917,240
- 1,834,975

**Comment by hanabi1224** [comment]
> Yes it would help to add those metrics, just to be sure. I did some testing and I can see capacity of `queries: FnvHashMap<QueryId, Query>` increasing like this over a period of 90 minutes:
> 
> * 457,443
> * 917,240
> * 1,834,975

@elmattic  Sure I can work on that, will update u once it's done

**Comment by hanabi1224** [comment]
@elmattic metrics updated, looks like it's still leaking `HashMap` space with my fix, will investigate more, maybe we can merge this PR first as it has completed its goal

https://github.com/ChainSafe/forest/actions/runs/3931143058/jobs/6722023668#step:13:8454
```
bitswap_container_capacities{type="BITSWAP_DB_QUERIES"} 448
bitswap_container_capacities{type="BITSWAP_REQUESTS"} 224
```

File: node/forest_libp2p/src/service.rs [hunk 2]
```diff
@@ -356,6 +375,41 @@ where
     }
 }
 
+async fn bitswap_timeout_task(
+    outgoing_bitswap_query_ids: BitswapOutgoingQueryTable,
+    outgoing_bitswap_query_cancellation_tx: Sender<libp2p_bitswap::QueryId>,
+) {
+    loop {
+        let mut timeout_queries = vec![];
```
Comment (elmattic) [inline node/forest_libp2p/src/service.rs:383]: Maybe just move outside the vector creation to avoid allocation?

**Comment by elmattic** [comment]
> @elmattic metrics updated, looks like it's still leaking `HashMap` space with my fix, will investigate more, maybe we can merge this PR first as it has completed its goal

Sure.

File: node/forest_libp2p/src/service.rs (hunk 2 above)
Comment (hanabi1224) [reply node/forest_libp2p/src/service.rs]: Fixed

**Pull request merged**